### ISSUE TITLE
Support for IDLv2 nullability semantics

### ISSUE THREAD

**Issue by sugmanue** [awslabs/smithy-rs#1767]
In order to be able to take full advantage of the new nullability semantics in IDLv2 the clients needs to have a mechanism to keep track whether or not a member with a default value has been updated, quoting from the [Defaults and Model Evolution](https://github.com/awslabs/smithy/blob/main/designs/defaults-and-model-evolution.md#default-value-serialization) document

> To allow servers to change default values if necessary, clients SHOULD NOT serialize default values unless the member is explicitly set to the default value. This implies that clients SHOULD implement a kind of "presence tracking" of defaulted members.

**Comment by jdisanti** [comment]
> I believe this is being tracked in another issue, and it is certainly on the roadmap to match v2 semantics in their entirety before a 1.0 release.

Yeah, there is a SDK issue for it: https://github.com/awslabs/aws-sdk-rust/issues/536

I'm going to leave this open to track implementation in smithy-rs.

**Comment by jdisanti** [comment]
I think we need to make some code generator changes to `StructureGenerator` and `BuilderGenerator` to fully support the new nullability semantics. I generated a new SDK with the latest IDLv2 SDK models and manually inspected the codegen diff looking for changes in optionality, and there was none.

I wrote a quick test in `StructureGeneratorTest` which fails:
```kotlin
    @Test
    fun `IDLv2 client @required fields are not options`() {
        val model = """
            namespace com.test
            structure MyStruct {
               @required
               foo: String,
               @required
               baz: Integer,
            }
        """.asSmithyModel(smithyVersion = "2.0")
        val struct = model.lookup<StructureShape>("com.test#MyStruct")

        val provider = testSymbolProvider(model)
        val project = TestWorkspace.testProject(provider)

        project.useShapeWriter(inner) { writer ->
            writer.withModule("model") {
                StructureGenerator(model, provider, writer, struct).render()
            }

            writer.rustBlock("fn check_optionality(s: &MyStruct)") {
                rust(
                    """
                    let _: &String = &s.foo;
                    let _: i64 = &s.baz;
                    """,
                )
            }
        }
        project.compileAndTest()
    }
```
The `SymbolVisitor` is coded up to use Smithy's `NullabilityIndex` to correctly determine if a shape should be optional or not, but it seems like something is getting lost in translation between that decision and the actual code generation for it.

**Comment by Velfi** [comment]
We're still waiting on some internal developments before we can begin work on this issue.

**Comment by jnicholls** [comment]
This appears to be assigned to Zelda now. Is this now scheduled for work on the smithy side? Really looking forward to this update, thank you!

**Comment by david-perez** [comment]
Regarding the interaction of `@input`/`@output` with structure member optionality, the below is something we should address when landing this.

[Structure member optionality](https://smithy.io/2.0/spec/aggregate-types.html#structure-optionality) says in the `@required` row, "Non-authoritative" column:

> Present unless also `@clientOptional` or part of an `@input` structure

Not all structures used as operation inputs are marked as `@input` in Smithy:

```smithy
operation MyOperation {
   input: SharedStruct
}

structure SharedStruct {
    @required
    member: String
}
```

So a smithy-rs user with the above model would hope that this would get synthesized as:

```rust
struct SharedStruct {
    member: String
}
```

However, we unconditionally apply the [`OperationNormalizer`](https://github.com/awslabs/smithy-rs/blob/main/codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/transformers/OperationNormalizer.kt) transform. This effectively adds a `MyOperationInput` structure shape to the model tagged with `@input`:

```smithy
operation MyOperation {
   input: MyOperationInput
}

@input
structure MyOperationInput {
    @required
    member: String
}
```

Hence the cell in that table says we should render this as:

```rust
struct MyOperationInput {
    member: Option<String>
}
```

And that will upset smithy-rs users, because from their perspective they rightfully think that they should get a non `Option`al field there.
